### PR TITLE
EVG-2425 Add Github Webhook configuration to Project Admin UI

### PR DESCRIPTION
**Pull request by richardsamuels**

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -418,7 +418,7 @@ func (settings *Settings) SessionFactory() *legacyDB.SessionFactory {
 
 func (s *Settings) GetGithubOauthToken() (string, error) {
 	token, ok := s.Credentials["github"]
-	if ok {
+	if ok || token == "" {
```
Comment (bsamek) [inline config.go]: We will never hit the error case.

File: public/static/js/projects.js [hunk 2]
```diff
@@ -171,6 +171,7 @@ mciModule.controller('ProjectCtrl', function($scope, $window, $http, $location)
           return v.alias + v.variant + v.task;
         });
         $scope.privateVars = data.ProjectVars.private_vars || {};
+        $scope.githubHookId = data.ProjectVars.github_hook_id || 0;
```
Comment (bsamek) [inline public/static/js/projects.js]: Should the id be a string?

File: service/project.go [hunk 3]
```diff
@@ -334,3 +362,76 @@ func (uis *UIServer) setRevision(w http.ResponseWriter, r *http.Request) {
...
 }
+
+func (uis *UIServer) setupGithubHook(projectRef *model.ProjectRef) (int, error) {
+	token, err := uis.Settings.GetGithubOauthToken()
+	if err != nil {
+		return 0, err
+	}
+
+	if uis.Settings.Api.GithubWebhookSecret == "" {
+		return 0, errors.New("config error")
```
Comment (bsamek) [inline service/project.go]: return a more explicit error

File: service/project.go [hunk 4]
```diff
@@ -334,3 +362,76 @@ func (uis *UIServer) setRevision(w http.ResponseWriter, r *http.Request) {
...
+}
+
+func (uis *UIServer) deleteGithubHook(projectRef *model.ProjectRef, hookID int) error {
+	token, err := uis.Settings.GetGithubOauthToken()
+	if err != nil {
+		return err
+	}
+
+	if uis.Settings.Api.GithubWebhookSecret == "" {
+		return errors.New("config error")
```
Comment (bsamek) [inline service/project.go]: more explicit error

File: service/project.go [hunk 5]
```diff
@@ -334,3 +362,76 @@ func (uis *UIServer) setRevision(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	httpClient, err := util.GetHttpClientForOauth2(token)
+	if err != nil {
+		return err
+	}
+	defer util.PutHttpClientForOauth2(httpClient)
+	client := github.NewClient(httpClient)
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Second)
```
Comment (bsamek) [inline service/project.go]: use `context.Background()` instead of `context.TODO()` (since you're actually using the context)

**Comment by richardsamuels** [comment]
Here's what it looks like when the user hasn't setup the github webhooks:
off
(Github patch definitions are hidden when they're not set up)

And here's what they look like when they're on:
on

File: service/project.go [hunk 6]
```diff
@@ -334,3 +362,76 @@ func (uis *UIServer) setRevision(w http.ResponseWriter, r *http.Request) {
...
+	defer cancel()
+	resp, err := client.Repositories.DeleteHook(ctx, projectRef.Owner, projectRef.Repo, hookID)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusNotFound {
```
Comment (bsamek) [inline service/project.go]: Should we return an error in the `http.StatusNotFound` case?

File: service/templates/projects.html [hunk 7]
```diff
@@ -276,15 +276,32 @@ <h3>Alerts</h3>
...
               <label class="distro-error">[[invalidKeyMessage]]</label>
             </div>
           </div>
         </div>
 
-        <div class="variables">
+        <div class="variables" ng-show="isSuperUser">
+          <div class="form-group">
+            <div class="col-header col-lg-6 form-control-static"> <h3> GitHub Webhook Configuration </h3>
+                <div class="muted small" ng-show="githubHookId === 0">Github Webhooks have not been enabled for this repository. In order to use Github Pull Request testing, tick the box below, and save the project. This will set up the webhook in this
```
Comment (bsamek) [inline service/templates/projects.html]: Downcase webhooks except in headlines.

File: config.go (hunk 1 above)
Comment (richardsamuels) [reply config.go]: If the token in the config is set to empty string, we will.

File: service/project.go (hunk 6 above)
Comment (richardsamuels) [reply service/project.go]: If the user accidentally deletes the hook from the repo admin page, not found can be returned if the user tries to disable and reenable the hook in the evergreen UI. So I chose to swallow this error here.

**Comment by tychoish** [comment]
lgtm modulo brian's comments.

File: public/static/js/projects.js (hunk 2 above)
Comment (richardsamuels) [reply public/static/js/projects.js]: Nope, it's the numeric hook id that github provides. If it's 0, the project is unconfigured. If not, it's configured

File: service/project.go (hunk 5 above)
Comment (richardsamuels) [reply service/project.go]: done

File: service/project.go (hunk 4 above)
Comment (richardsamuels) [reply service/project.go]: done

File: service/project.go (hunk 3 above)
Comment (richardsamuels) [reply service/project.go]: done

File: service/templates/projects.html (hunk 7 above)
Comment (richardsamuels) [reply service/templates/projects.html]: done

**Pull request merged**